### PR TITLE
Web3: Fix TX message compilation accounting

### LINKED ISSUE #10897: web3.js: Conflicting code / test logic for Transaction serialization
#### Problem

While fixing a bug in transaction message serialization, I happened upon the following paradox...

We have a test:
https://github.com/solana-labs/solana/blob/3884323cb6333d3f085624a008b5fac14dc5fe4c/web3.js/test/transaction.test.js#L198-L200
Which is named and has a golden value to imply it is testing serialization of a transaction that has not yet been signed.

However!  `Transaction.serialize()` starts like this:
https://github.com/solana-labs/solana/blob/3884323cb6333d3f085624a008b5fac14dc5fe4c/web3.js/src/transaction.js#L395-L406
Implying that attempting to serialize an unsigned transaction should throw.

#### Proposed solution

I can't think of a compelling reason to serialize an unsigned transaction, where just the message wouldn't do the same job.  So I'm thinking the RightThing(TM) to do here is replace the test with one that ensures `Transaction.serialize()` throws on an unsigned transaction and another that tests `Transaction.serializeMessage()`

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

The web3.js `Transaction.compileMessage()` method's accounting of `readonlySignedAccounts` depends upon the initial contents of the `signatures` field. Since this method _also_ pre-allocates signature entries, calling it twice will return different message headers.

#### Summary of Changes

- Switch external signer test to use a `DelegateStake` rather than `Transfer` instruction, which exhibits this bug
- Bring message compilation logic in-line with Rust implementation
- Logically align code, comments and test for serializing unsigned transactions

Fixes #10897

### CODE REVIEW

**Comment by t-nelson** [comment]
What's the trick to getting these timed out RPC tests to pass CI?

**Comment by CriesofCarrots** [comment]
> What's the trick to getting these timed out RPC tests to pass CI?

@t-nelson , those live tests fail locally, so I think there's actually a broken test. I haven't looked to carefully at your code changes yet, but the failure seems to be in transaction-payer.test.js, which does seem related to your changes...

File: web3.js/src/transaction.js [hunk 1]
```diff
@@ -210,54 +210,79 @@ export class Transaction {
...
+    // Append programID account metas
+    programIds.forEach(programId => {
+      accountMetas.push({
+        pubkey: programId,
+        isSigner: false,
+        isWritable: false,
+      });
+    });
+
+    // Prefix accountMetas with feePayer here whenever that gets implemented
```
Comment (CriesofCarrots) [inline web3.js/src/transaction.js:235]: The lack of this is your issue.

Need something like this, I imagine:
```
this.signatures.forEach(signature => {
      if (!accountMetas.find(meta => meta.pubkey.toString() === signature.publicKey.toString())) {
        accountMetas.unshift({
          pubkey: signature.publicKey,
          isSigner: true,
          isWritable: true,
        });
      }
    });
```

**Comment by jstarry** [comment]
FYI there was some CI flakiness that I observed recently and I patched it up here: https://github.com/solana-labs/solana/pull/10956

**Comment by garious** [comment]
Instead of putting time here, can we compile the Rust Message library to WebAssembly?

File: web3.js/src/transaction.js [hunk 2]
```diff
@@ -210,54 +210,79 @@ export class Transaction {
     let numReadonlySignedAccounts = 0;
     let numReadonlyUnsignedAccounts = 0;
 
-    const accountKeys = this.signatures.map(({publicKey}) =>
-      publicKey.toString(),
-    );
-    const programIds: string[] = [];
+    const programIds: PublicKey[] = [];
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply web3.js/src/transaction.js:213]: So it does :slightly_smiling_face: Will fix

File: web3.js/src/transaction.js (hunk 1 above)
Comment (t-nelson) [reply web3.js/src/transaction.js:235]: Nice catch! :pray:

**Comment by t-nelson** [comment]
> Instead of putting time here, can we compile the Rust Message library to WebAssembly?

This isthe result of some work I did last week with a third party wallet provider to unblock their staking integration.

But it would indeed be nice to not have this code to maintain :wink:

File: web3.js/src/transaction.js (hunk 1 above)
Comment (t-nelson) [reply web3.js/src/transaction.js:235]: I went with https://github.com/solana-labs/solana/pull/10950/commits/1e11573802be58036527c29bf40adf79221d2b15#diff-86eea70a26ecf01c3229f804beb21ac4R260-R272

File: web3.js/src/transaction.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-nelson) [reply web3.js/src/transaction.js]: Back to strings!  I figured there was a reason for all of those conversions

**Pull request merged**